### DISCUSSION TITLE
Java EE to Spring 3.x.x

### DISCUSSION THREAD

**Discussion by bohdant-of** [openrewrite/rewrite#4475] [Q&A] [answered]
Hi OpenRewrite Community,

I would like to know if there is a streamlined (preferably documented) way to migrate from Java EE to Spring Boot 3.x.x using OpenRewrite. Cheers!

**Comment by timtebeek** [comment, marked as answer]
There's no single recipe for Java EE → Spring Boot 3.x, but the combination of these gets you most of the way:

- **[rewrite-migrate-java](https://github.com/openrewrite/rewrite-migrate-java)** — Jakarta EE migration recipes (javax → jakarta namespace, etc.)
- **[rewrite-spring](https://github.com/openrewrite/rewrite-spring)** — Spring Boot 2.x → 3.x migration recipes

You'd typically run the Jakarta migration first, then apply the Spring Boot 3.x upgrade recipes. See the [Spring Boot 3.x migration guide](https://docs.openrewrite.org/recipes/java/spring/boot3/springboot3bestpractices) on the OpenRewrite docs for the recommended composite recipe.
